### PR TITLE
Add the original data to attributes and allow to access it in ExpressionLanguage

### PR DESCRIPTION
**Pull request by antograssiot**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1646
| License       | MIT
| Doc PR        | TODO

Alternative approach to #2710 as discussed there.

The good thing about this one is that it is totally BC.
The disadvantage is that it might be harder to discover even if covered by documentation and users will not be "protected" out of the box.

### CODE REVIEW

File: src/EventListener/ReadListener.php [hunk 1]
```diff
@@ -107,5 +107,6 @@ public function onKernelRequest(GetResponseEvent $event): void
         }
 
         $request->attributes->set('data', $data);
+        $request->attributes->set('originalData', is_object($data) ? clone $data : $data);
```
Comment (dunglas) [inline src/EventListener/ReadListener.php]: Should be `origina_data` according to Symfony convention IIRC.

File: tests/Fixtures/TestBundle/Document/SecuredDummy.php [hunk 2]
```diff
@@ -30,7 +30,8 @@
  *         "post"={"access_control"="has_role('ROLE_ADMIN')"}
  *     },
  *     itemOperations={
- *         "get"={"access_control"="has_role('ROLE_USER') and object.getOwner() == user"}
+ *         "get"={"access_control"="has_role('ROLE_USER') and object.getOwner() == user"},
+ *         "put"={"access_control"="has_role('ROLE_USER') and original_object.getOwner() == user"},
```
Comment (dunglas) [inline tests/Fixtures/TestBundle/Document/SecuredDummy.php]: May I suggest to test both?

```suggestion
 *         "put"={"access_control"="has_role('ROLE_USER') and object.getOwner() == user and  original_object.getOwner() == user"},
```

**Comment by dunglas** [comment]
Great one! I'm just a bit unsure about the naming. React use `prevState`, what about `previous_data`?

File: tests/Fixtures/TestBundle/Document/SecuredDummy.php (hunk 2 above)
Comment (antograssiot) [reply tests/Fixtures/TestBundle/Document/SecuredDummy.php]: Testing both isn't ok @dunglas as we are trying to make sure that only the actual owner can update the owner to become someone else. This someone shouldn't be be able to assign the object to himself

Comment (antograssiot) [reply tests/Fixtures/TestBundle/Document/SecuredDummy.php]: Well I can check owner for the "get" operation if you prefer, they are the same in this case...

**Comment by teohhanhui** [comment]
@antograssiot What happened with your last force push? Looks like it's using the wrong branch. This should target `master`.

**Comment by antograssiot** [comment]
@teohhanhui it already targets master and the diff looks good to me 🤔
What does looks wrong to you ?

**Comment by teohhanhui** [comment]
@antograssiot Oh, never mind. I was confused by the force push diff. :laughing:

**Comment by teohhanhui** [comment]
Thanks @antograssiot! :tada:

**Comment by antograssiot** [comment]
I'll send a doc PR

**Pull request merged**